static/js/containers/CourseSearchPage.js

```diff
@@ -121,10 +121,45 @@ export class CourseSearchPage extends React.Component<Props, State> {
         ["platform", []],
         ["availability", []],
         ["topics", []]
-      ])
+      ]),
+      currentFacetGroup: null
     })
   }
 
+  mergeFacetOptions = (group: string) => {
+    const { facets } = this.props
+    const { activeFacets, currentFacetGroup } = this.state
+    const resultFacets = facets ? facets.get(group) : null
+
+    // add any selected facets not in results to the group
+    if (resultFacets && _.has(resultFacets, "buckets")) {
+      // $FlowFixMe: this is not undefined
+      activeFacets.get(group).map(facet => {
+        if (!_.find(resultFacets.buckets, { key: facet })) {
+          resultFacets.buckets.push({
+            key:       facet,
+            doc_count: 0
+          })
+        }
+      })
+    }
+
+    // if this is the currentFacetGroup, add any new facets in that group from search results
+    if (currentFacetGroup && currentFacetGroup.keys().next().value === group) {
+      const mergedFacetGroup = _.clone(currentFacetGroup).get(group)
+      if (mergedFacetGroup && resultFacets && _.has(resultFacets, "buckets")) {
+        resultFacets.buckets.map(facet => {
+          if (!_.find(mergedFacetGroup.buckets, { key: facet.key })) {
+            mergedFacetGroup.buckets.push(facet)
+          }
+        })
+      }
+      return mergedFacetGroup
+    }
+
+    return resultFacets
+  }
```

Comment:
this logic to store the facet options in the `currentFacetGroup` value and merge with the changed facets on `this.props` strikes me overall as pretty complicated and hard to follow. Do you think there are any opportunities to simplify it? I was wondering in particular about a utility function for merging these `FacetResults` objects. If you had a `mergeFacets` function which did that work for you, then I think this function could look something like:


```js
if (!facets) {
  return null
}

const emptyActiveFacets = { buckets: activeFacets.map(facet => ({ key: facet, doc_count: 0 })) }

if (currentFacetGroup.get(group)) {
  return mergeFacetResults(
    facets.get(group),
    emptyActiveFacets,
    currentFacetGroup.get(group)
  )
} else {
  return mergeFacetResults(
    facets.get(group),
    emptyActiveFacets
  )
}
```
 
IMO it would be easier to read and follow the code this way, and there wouldn't be this duplicated logic to merge the new result facets with the active facets and the 'current' facets.

I think a merge function could look something like this:

```js
const mergeFacetResults = (...args: Array<FacetResult>) => ({
  buckets: args
    .map(R.prop("buckets"))
    .reduce((buckets, acc) => R.unionWith(R.eqBy(R.prop("key")), buckets, acc))
})
```
assuming I've understood the data structure :smile: